syntax = "proto3";
// import "google/protobuf/any.proto";
// import "google/protobuf/wrappers.proto";
import "google/protobuf/descriptor.proto";

package nanorpc;

extend google.protobuf.MessageOptions {
	// Specifies that message is a method argument list.
	bool expand_as_arguments = 90001;

	// Specifies that the message is an enumeration wrapper.
	bool enum_wrapper = 90002;
}

// extend google.protobuf.FieldOptions {
// }

extend google.protobuf.ServiceOptions {
	// Specifies that the service is an event source (implemented on client).
	bool event_source = 90301;
}

extend google.protobuf.MethodOptions {
	// Specifies that the method is actually a description of property.
	bool is_property = 90401;

	// Specifies that the method is asynchronous.
	bool async = 90402;
}


enum RpcStatus {
	RpcSucceeded = 0;
	RpcChannelFailure = 1;
	RpcUnknownMethod = 2;
	RpcProtocolError = 3;
	RpcUnknownInterface = 4;
	RpcInvalidCallParameter = 5;
}

message RpcCall {
	string service = 1;
	string method = 2;
	uint32 object_id = 3;
	bytes call_data = 4;
}

message RpcResult {
	RpcStatus status = 1;
	string error_message = 2;
	bytes result_data = 3;
}

message RpcMessage {
	int32 id = 1;
	RpcCall call = 2;
	RpcResult result = 3;
}

message RpcVoid {
}

message WideStringValue {
	string value = 1;
}

message SInt32Value {
	sint32 value = 1;
}

message SInt64Value {
	sint64 value = 1;
}

message RpcObject {
	uint32 object_id = 1;
}

message RpcEvent {
	string event_name = 1;
}